src/encode.rs

```diff
@@ -0,0 +1,314 @@
+
+        // Encode primes
+        blocks.extend(
+            self.primes
+                .iter()
+                .map(|p| ASN1Block::Integer(0, to_bigint(p))),
+        );
+        to_der(&ASN1Block::Sequence(0, blocks)).map_err(|e| Error::EncodeError {
+            reason: format!("failed to encode ASN.1 sequence of blocks: {}", e),
+        })
+    }
+
+    fn to_pkcs8(&self) -> Result<Vec<u8>> {
+        let version = ASN1Block::Integer(0, to_bigint(&BigUint::zero()));
+        let oid = ASN1Block::ObjectIdentifier(0, rsa_oid());
+        let alg = ASN1Block::Sequence(0, vec![oid]);
+        let octet_string = ASN1Block::OctetString(0, self.to_pkcs1()?);
+        let blocks = vec![version, alg, octet_string];
+
+        to_der(&ASN1Block::Sequence(0, blocks)).map_err(|e| Error::EncodeError {
+            reason: format!("failed to encode ASN.1 sequence of blocks: {}", e),
+        })
+    }
+}
+
+pub trait PublicKeyEncoding: PublicKey {
+    /// Encodes an RSA Public key to into `PKCS1` bytes.
+    ///
+    /// This data will be `base64` encoded which would be used
+    /// following a `-----BEGIN RSA PUBLIC KEY-----` header.
+    ///
+    /// <https://tls.mbed.org/kb/cryptography/asn1-key-structures-in-der-and-pem>
+    fn to_pkcs1(&self) -> Result<Vec<u8>> {
+        let n = ASN1Block::Integer(0, to_bigint(&self.n()));
+        let e = ASN1Block::Integer(0, to_bigint(&self.e()));
+        let blocks = vec![n, e];
+
+        to_der(&ASN1Block::Sequence(0, blocks)).map_err(|e| Error::EncodeError {
+            reason: format!("failed to encode ASN.1 sequence of blocks: {}", e),
+        })
+    }
+    /// Encodes an RSA Public key to into `PKCS8` bytes.
+    ///
+    /// This data will be `base64` encoded which would be used
+    /// following a `-----BEGIN PUBLIC KEY-----` header.
+    ///
+    /// <https://tls.mbed.org/kb/cryptography/asn1-key-structures-in-der-and-pem>
+    fn to_pkcs8(&self) -> Result<Vec<u8>> {
+        let oid = ASN1Block::ObjectIdentifier(0, rsa_oid());
+        let alg = ASN1Block::Sequence(0, vec![oid]);
+
+        let bz = self.to_pkcs1()?;
+        let octet_string = ASN1Block::BitString(0, bz.len(), bz);
+        let blocks = vec![alg, octet_string];
+
+        to_der(&ASN1Block::Sequence(0, blocks)).map_err(|e| Error::EncodeError {
+            reason: format!("failed to encode ASN.1 sequence of blocks: {}", e),
+        })
+    }
+}
+
+impl PublicKeyEncoding for RSAPublicKey {}
+
+#[cfg(all(test, feature = "pem"))]
+mod tests {
+    use super::{PrivateKeyPemEncoding, PublicKeyPemEncoding};
+    use crate::{RSAPrivateKey, RSAPublicKey};
+    use rand::thread_rng;
+    use rand::SeedableRng;
+    use rand_xorshift::XorShiftRng;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn priv_pem_encoding_pkcs8() {
+        const PKCS8_PRIVATE_KEY: &str = "-----BEGIN PRIVATE KEY-----\r\nMEACAQAwCwYJKoZIhvcNAQEBBC4wLAIBAAIJAJGyCM1NTAwDAgMBAAECCQCMDHwC\r\nEdIqAQIFAMEBAQECBQDBQAkD\r\n-----END PRIVATE KEY-----\r\n";
```

Comment:
It's strange that this contains `\r` in the output. Maybe the `to_pem_pkcs[1|8]` methods should return a `Pem` struct and the user can call `pem::encode` or `pem::encode_config` themselves so that they have control over the line endings.

Reply:
hmm yeah that is weird that is the default, or could have the `EncodeConfig` exposed and just have the function take in an optional param of this? seems a bit more ergonomic than requiring importing pem and doing manually

Reply:
Don't Windows machines handle LF-only line endings these days? Why not just make that the default?

Worst case we can see if someone complains.

Reply:
My guess is that `encode` detects the platform and makes the decision for the user. I'm primarily on Windows, but I **never** want `CRLF` line endings.  

>Don't Windows machines handle LF-only line endings these days? Why not just make that the default?

I personally think that `CRLF` is a vestige that should be purged from modern computing :) 

However, there are some examples where _only_ allowing `LF` endings could be problematic. One such example is when people set `core.autocrlf=true` in their git config. Git will try to inject `CR` in your source but remove it when committing. (Hint: `core.autocrlf=inline` is the way and the light and will use `LF` on windows). 

Deciding either way is opinionated and potentially problematic. I'm sure there are examples other than the one I provided. Returning the `Pem` struct is un-opinionated, flexible, and gives the user full control.

> seems a bit more ergonomic than requiring importing pem and doing manually

The `pem` crate is re-exported at the top level. They'll need to use `pem` or `pem::EncodeConfig` either way. A single extra line isn't isn't that un-ergonomic IMO.

```rust
let pem = private_key.to_pem_pkcs1()?;
let pem_str = rsa::pem::encode(&pem); // or encode_config
```

Or maybe take an `Option<&EncodeConfig>` and call `encode` if `None` is provided. I still think the former is the most flexible but this isn't too bad either:
```rust
let pem_str = private_key.to_pem_pkcs1(None)?;
```

Reply:
> However, there are some examples where only allowing LF endings could be problematic. One such example is when people set core.autocrlf=true in their git config

Seems like that would only impact users checking private keys into git? I'm not sure that's a use case worth caring about 😉

Reply:
> Seems like that would only impact users checking private keys into git? I'm not sure that's a use case worth caring about 

You can put a password on the key, but I'm not advocating putting keys into git! :) I'm mostly just trying to express that there are valid use cases, even if we aren't necessarily aware of them.

I don't think being opinionated here outweighs a single extra line or having the function take an `Option<&EncodeConfig>`. 

The more I think about it, having a signature of `to_pem_pkcs1(&self, Option<&EncodeConfig>)` definitely feels like the best of both worlds - ergonomic, but flexible if needed.

```rust
let pem_str = private_key.to_pem_pkcs1(None)?;
```

Reply:
I don't have an opinion (just did this for fun) so let me know what route is preferred when you guys decide. That way (which is what I was suggesting) just made sense to me. Also I guess what None implies can be opinionated too, so if that way is chosen let me know if you want the default to be LF endings and I'll do this when I have free time

Reply:
> Also I guess what None implies can be opinionated too

To me it's semantically unclear in that context what `None` means

Reply:
+1 for `Default::default` instead of `None`

Reply:
though I like having `to_pem_pkcs1` being an alias for `to_pem_pkcs1_with_config(Default::default)`

Reply:
and have to_pem_pkcs{1,8} use the pem default (CRLF) or use LF? I only ask because there seemed to be opinions against CRLF

Reply:
I think there are systems that reject CRLF line endings (let me look into that).

I have no idea why it's the default. It seems wrong to me. LF should be the default...

Reply:
`EncodeConfig` doesn't implement default, which is why I suggested using `Option`.

Also, `encode` accepts the config as a reference. I think that `to_pem_pkcs1_with_config` should also take `&EncodeConfig` by reference.

Reply:
You could always use a newtype which exposes the options we care about. It seems like this is the only one?

Reply:
updated. I just used the pem `EncodeConfig` type, but happy to switch over if that is what you guys want.

> Also, encode accepts the config as a reference. I think that to_pem_pkcs1_with_config should also take &EncodeConfig by reference.

It doesn't? The type is moved in and is copy, but happy to switch to passing in a reference if the previous change is made to future proof if the config type gets bulkier (but I assume that would be a warning because it's just an enum without data rn)

Also made the default LF, but that can be switched very easily also

Reply:
> It doesn't? The type is moved in and is copy

I double checked and you are correct. I could have sworn it was passed by reference, but I stand corrected.

Reply:
Ah, it was passed by reference in an earlier version:

https://docs.rs/pem/0.6.1/pem/fn.encode_config.html